File: packages/mlmpfr.4.1.0/testsuite/specialfunctions.c
/* This file is part of mlmpfr.

  mlmpfr is free software: you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  mlmpfr is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with mlmpfr. If not, see
 <http://www.gnu.org/licenses/>. */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <mpfr.h>

char *rounding_to_string (int r)
{
  char * buf = malloc (32);
  if (r == 0)
    {
      strcpy (buf, "Correct");
    }
  else if ( r < 0)
    {
      strcpy (buf, "Lower");
    }
  else
    {
      strcpy (buf, "Greater");
    }
  return buf;
}

void all (mpfr_t op1, mpfr_t op2)
{
  int ter;
  mpfr_t rop, sop, op22;
  mpfr_init (rop);
  mpfr_init (sop);
  mpfr_init (op22);

  ter = mpfr_log (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  if (mpfr_sgn (op1) >= 0)
    {
      ter = mpfr_log_ui (rop, (mpfr_get_si (op1, MPFR_RNDN)), MPFR_RNDN);
      mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
    }
  else
    {
      mpfr_set_nan (rop);
      mpfr_printf ("%.Re None\n", rop);
    }

  ter = mpfr_log2 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_log10 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_exp (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_exp2 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_exp10 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_sin (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_cos (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_tan (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_sin_cos (rop, sop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %.Re %s\n", rop, sop, ter == 0 ? rounding_to_string (ter) : "None");

  ter = mpfr_sec (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_csc (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_cot (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_acos (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_asin (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_atan (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_atan2 (rop, op1, op2, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_sinh (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_cosh (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_tanh (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_sinh_cosh (rop, sop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %.Re %s\n", rop, sop, ter == 0 ? rounding_to_string (ter) : "None");

  ter = mpfr_sech (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_csch (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_coth (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_acosh (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_asinh (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_atanh (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  if (mpfr_sgn (op1) >= 0)
    {
      ter = mpfr_fac_ui (rop, (mpfr_get_si (op1, MPFR_RNDN)), MPFR_RNDN);
      mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
    }
  else
    {
      mpfr_set_nan (rop);
      mpfr_printf ("%.Re None\n", rop);
    }

  ter = mpfr_log1p (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_expm1 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_eint (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_eint (rop, op2, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_li2 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_gamma (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_gamma_inc (rop, op1, op2, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_lngamma (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  int sign;
  ter = mpfr_lgamma (rop, &sign, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s %s\n", rop, sign > 0 ? "Positive" : "Negative", rounding_to_string (ter));

  ter = mpfr_digamma (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_zeta (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_beta (rop, op1, op2, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_erf (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_erfc (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_j0 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_j1 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  mpfr_abs (op22, op2, MPFR_RNDN);
  ter = mpfr_jn (rop, mpfr_get_ui (op22, MPFR_RNDN), op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_y0 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_y1 (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_yn (rop, mpfr_get_ui (op22, MPFR_RNDN), op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_fma (rop, op1, op2, rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_fms (rop, op1, op2, rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_fmma (rop, op1, op2, rop, rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_fmms (rop, op1, op2, rop, rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_agm (rop, op1, op2, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_hypot (rop, op1, op2, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  ter = mpfr_ai (rop, op1, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_const_log2 (rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_const_pi (rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_const_euler (rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));
  ter = mpfr_const_catalan (rop, MPFR_RNDN);
  mpfr_printf ("%.Re %s\n", rop, rounding_to_string (ter));

  int size = 1000000;
  mpfr_ptr ptab [size];
  int i;
  for (i = 0; i < size; i+=2)
    {
      ptab[i] = op1;
      ptab[i+1] = op2;
    }
  ter = mpfr_sum (rop, ptab, size, MPFR_RNDN);
  mpfr_printf ("%.Re %s %d\n", rop, rounding_to_string (ter), size);
  ter = mpfr_dot (rop, ptab, ptab, size/10, MPFR_RNDN);
  mpfr_printf ("%.Re %s %d\n", rop, rounding_to_string (ter), size/10);

  mpfr_clear (rop);
  mpfr_clear (sop);
  mpfr_clear (op22);
}

int main ()
{
  mpfr_t op1, op2;
  mpfr_inits (op1, op2, NULL);

  mpfr_set_d (op1, 1. / 3, MPFR_RNDN);
  mpfr_set_d (op2, 1. / 10, MPFR_RNDN);

  all(op1, op2);
  printf("\n");

  mpfr_set_d (op1, -4. / 3, MPFR_RNDN);

  all(op1, op2);
  printf("\n");

  mpfr_set_d (op1, 1. / 3, MPFR_RNDN);
  mpfr_set_d (op2, -113. / 10, MPFR_RNDN);

  all(op1, op2);
  printf("\n");

  mpfr_set_d (op1, +0, MPFR_RNDN);
  mpfr_set_d (op2, -1. / 10, MPFR_RNDN);

  all(op1, op2);
  printf("\n");

  mpfr_set_d (op1, -0., MPFR_RNDN);
  mpfr_set_d (op2, -0., MPFR_RNDN);

  all(op1, op2);

  mpfr_clears (op1, op2, NULL);
  return 0;
}

File: packages/mlmpfr.4.1.0/testsuite/specialfunctions.ml
(* This file is part of mlmpfr.

  mlmpfr is free software: you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  mlmpfr is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with mlmpfr. If not, see
  <http://www.gnu.org/licenses/>. *)

open Printf
open Common
module M = Mpfr

let all op1 op2 =
  let r = M.log op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r =
    try
      M.log_int (M.get_int op1)
    with M.Invalid_integer_input _ ->
      M.make_nan ()
  in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.log2 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.log10 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.exp op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.exp2 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.exp10 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.sin op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.cos op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.tan op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r, s = M.sin_cos op1 in printf "%s %s %s\n" (M.get_formatted_str r) (M.get_formatted_str s) (rounding_to_string r);
  let r = M.sec op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.csc op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.cot op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.acos op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.asin op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.atan op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.atan2 op1 op2 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.sinh op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.cosh op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.tanh op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r, s = M.sinh_cosh op1 in printf "%s %s %s\n" (M.get_formatted_str r) (M.get_formatted_str s) (rounding_to_string r);
  let r = M.sech op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.csch op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.coth op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.acosh op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.asinh op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.atanh op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r =
    try
      M.fac_int (M.get_int op1)
    with M.Invalid_integer_input _ ->
      M.make_nan ()
  in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.log1p op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.expm1 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.eint op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.eint op2 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.li2 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.gamma op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.gamma_inc op1 op2 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.lngamma op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r, s = M.lgamma op1 in printf "%s %s %s\n" (M.get_formatted_str r)
                                    (match s with M.Positive -> "Positive" | M.Negative -> "Negative") (rounding_to_string r);
  let r = M.digamma op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.zeta op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.beta op1 op2 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.erf op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.erfc op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.j0 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.j1 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.jn (M.get_int (M.abs op2)) op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.y0 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.y1 op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.yn (M.get_int (M.abs op2)) op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.fma op1 op2 r in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.fms op1 op2 r in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.fmma op1 op2 r r in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.fmms op1 op2 r r in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.agm op1 op2 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.hypot op1 op2 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.ai op1 in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.const_log2 (M.get_default_prec ()) in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.const_pi (M.get_default_prec ()) in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.const_euler (M.get_default_prec ()) in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let r = M.const_catalan (M.get_default_prec ()) in printf "%s %s\n" (M.get_formatted_str r) (rounding_to_string r);
  let rec s ?acc:(acc = []) n =
    if n = 0 then acc
    else s ~acc:(op1::op2::acc) (n-2)
  in
  let l = (s 1000000) in
  let r = M.sum l in
  printf "%s %s %d\n" (M.get_formatted_str r) (rounding_to_string r) (List.length l);
  let m = (s 100000) in
  let r = M.dot l m in
  printf "%s %s %d\n" (M.get_formatted_str r) (rounding_to_string r) (List.length m)

let _ =
  all (M.make_from_float (1. /. 3.)) (M.make_from_float (1. /. 10.)); printf "\n";
  all (M.make_from_float ((~-. 4.) /. 3.)) (M.make_from_float (1. /. 10.)); printf "\n";
  all (M.make_from_float (1. /. 3.)) (M.make_from_float ((~-. 113.) /. 10.)); printf "\n";
  all (M.make_zero M.Positive) (M.make_from_float ((~-. 1.) /. 10.)); printf "\n";
  all (M.make_zero M.Negative) (M.make_zero M.Negative);
  Gc.full_major (); (* garbage collector full major *)
